signalmetrics/signal
# Data pipeline

## Where I think we are
- Getting data from the browser into the database should be superficially working. 
- There will be bugs that need to be ironed out, but they'll be easier to see as we're integrating data into the app.
- Locally, I can run `SignalEvent::query()->where('title', 'Some Title')->count();` and it works!
- At this moment, I'm "pencils down" until we chat!



## Detailed notes

### Test
The PEST test that's in there passes for me. Does it pass for you too?

--

### How to install in a test repo
1. Require package in a local repo
2. `php artisan signal:install` (you should then have signal.sqlite with migrations run)
3. add `@signal` to a blade file in the repo where you installed the package
4. visit a page on the site.
5. you should see an event in signal_today.

--

### Data flow
**This may be easier to talk through, so don't feel like you need to digest all of this! It's just here for reference later so I remember.**
- `signal.js` grabs a bunch of data from the page and sends to IngestionController
- IngestionController decides if it's a page_view or event. (right now, event does nothing)
- page_views go through a whole data processing pipeline.
- once processed, the page_view gets stored in `today`
   - they go into `today` because there are no indexes, so it's super fast.
   - Also, if the `events` table was changing every 30 seconds, it'd be annoying.
 - `MoveDataToEventsCommand.php` automatically moves data from `today` into `events` every 30 minutes
